### ISSUE TITLE
Vote reminder does not work as intended

### ISSUE THREAD

**Issue by mrcoffee1026** [teozfrank/UltimateVotes#18]
For vote reminder:

```
    votereminder:
      enabled: true
      interval: 86400

```

What unit of time is the interval?? Is it even time?  Seconds? Ticks? Some other unity that is somehow completely unrelated to time?  I can't seem to get these to "remind" in a reasonable manner.  It either reminds like every 20 minutes whether you've voted or not... or it starts reminding like 8 hours past the time I can start voting again (with most sites that's 24 hours).  Of course I'm just blindly stabbing at the values since I have no idea what I'm specifying (seconds, minutes, number of big macs one can consume in a minute, etc.).  Please specify what this actually counts... and does the timer begin from the time the last vote is received from that player (very useful)?  Or does it start counting from the moment the player logs in (virtually useless)? 

Thanks!

**Comment by teozfrank** [comment]
Its time in ticks. Multiply the time by 20 to get the value in seconds.

> and does the timer begin from the time the last vote is received from that player (very useful)? Or does it start counting from the moment the player logs in (virtually useless)?

No it is based on the date the player votes on if it is greater than 1 day then they are reminded to vote again. The time you are setting is how often the server reminds the players who have not voted and are online to vote. Does that make sense?

**Comment by mrcoffee1026** [comment]
This does not have the expected effect then?  "Greater than one"... would mean it wouldn't start providing reminders until 2 days have elapsed (if yesterday was 4/1, reminders will begin on 4/3). I think you mean greater than or equal to one, since that's what I've observed... however... that means reminder ticks tart at 12:00am on 4/2, regardless of what time the user voted last.  Do you think it would be possible to... instead of starting it at this day difference by one... that every vote could simply update a "last_time_voted" timestamp... and then the reminders could just begin at whatever value we put in the config file from that point in time instead?  As THIS is what I thought this setting would have done.  your explanation does explain why it does not work this way currently , but I believe this would be the desired way of doing this for anyone changing this setting.

Thanks for jumping back on these things, I really appreciate it.

**Comment by teozfrank** [comment]
Yes but that would require a database change, the lastvoted is a date not a datetime, I'll look into it.

**Comment by mrcoffee1026** [comment]
If that's what's needed then that's what I'd want.  At least then the reminders would be genuinely useful, and i could set them to remind people right at 24 hours or maybe 23.5 hours as there are plenty of sites that only allow one vote per day on the same IP or for the same username.  Having the timer/reminders start at 12am it results in fairly useless reminders... either hours before a user can vote again, or hours after they should have already done so if they don't want to completely miss a day of voting.  Anyway if you make any progress on this my players and I would be most appreciative.

**Comment by teozfrank** [comment]
This has been fixed in a dev release on discord, it will now remind a player to vote 24 hours from when they voted using the new votelog table which uses timestamps and not dates. UltimateVotes_2.0.1_SNAPSHOT-79d4054

**Comment by mrcoffee1026** [comment]
My initial attempt with this release so far has not shown any deviation from previous behavior:  every reminder is sent to every person on the server at exactly the same time regardless of when they last voted.  I'm hoping this is just a glitch due to this being the first time run with the option on, but first or second time shouldn't make a difference, it should be using last voted time as the point at which the reminder clock gets set and issued.  If it's unable to do that, then the reminders serve no purpose other than to irritate players who still have 12-24 hours to wait before they can vote at any of those sites again.

**Comment by teozfrank** [comment]
> 
> 
> My initial attempt with this release so far has not shown any deviation from previous behavior: every reminder is sent to every person on the server at exactly the same time regardless of when they last voted. I'm hoping this is just a glitch due to this being the first time run with the option on, but first or second time shouldn't make a difference, it should be using last voted time as the point at which the reminder clock gets set and issued. If it's unable to do that, then the reminders serve no purpose other than to irritate players who still have 12-24 hours to wait before they can vote at any of those sites again.

What you are describing is not what is actually going on internally in the plugin. I understand what you are looking for so in fact you will see simular behaviour in how the vote reminder works but the determination of if a player has voted or not is what has actually changed.

Let me descrbe the change so you can understand what is needed to test it.

1. When a player joins, they are either added to a voted list or a not voted list.

2. To check if the player has voted, the plugin checks the votelog and checks has the player voted in the last 24 hours. If the player has voted they are placed in the has voted list. If the player has not voted they are placed in the not voted list. 

3. Then every X amount of time the plugin loops through the list of not voted players and send them a vote reminder. Thats how it works.

Let me be clear this is NOT an official release its an UNTESTED BUILD where I made the code changes and did not test it as I did not have the time. Thats what SNAPSHOT jars are usually, so if you do not see a change or if you see incorrect behavour yes it could happen, but I need YOUR help in determining that OR wait until I have gotten some time to verify it myself.

The best way to test this would be:
1. enable debug mode on the server where you are joining.
2. Join the server, it should say not voted.
3. Leave the server
4. Vote for the server
5. Join the server
6. It should say has voted.

If 6. Does not work then you would be right and there is still an outstanding problem to which I would need to debug further.

**Comment by mrcoffee1026** [comment]
Oh ok, so it's still just going to bulk remind everyone on the server at the same time, it just may not if they've voted within the last 24 hours... firstly, it's just reminding everyone whether they voted or not.  Secondly, this sort of reminder is not an improvement on the one before since it's just doing them all at the same time regardless of when they voted last.  The reminder should be per player from the time they voted last not from the time the server started or whatever it's doing now.  Since it's not doing that, we won't want this feature enabled since it's just irritating to people.  They'll get a reminder long before they can use the vote sites again... Vote sites have timers per player and restrict people from voting more than once per 12 or 24 hours.  It does not make sense for this server to send them a reminder before that time has elapsed.

I would do the debug testing but the results won't matter since based on your above description, this feature won't work as desired anyway even if it is functioning perfectly.

**Comment by teozfrank** [comment]
> Oh ok, so it's still just going to bulk remind everyone on the server at the same time, it just may not if they've voted within the last 24 hours

Its not a "may", it should not.

> firstly, it's just reminding everyone whether they voted or not. 

It shouldn't do that.

> Secondly, this sort of reminder is not an improvement on the one before since it's just doing them all at the same time regardless of when they voted last.

Again it shouldn't do that.

> The reminder should be per player from the time they voted last not from the time the server started or whatever it's doing now

I will consider another approach if required. The plugin does not use any time when the server started it is purely driven from the timestamp the player last voted which is recorded in the database.

I will investigate further as the change has not fixed it by the sounds of things. My apologies for the clear frustration this has caused. I would advise to just disable the feature until I have fixed it, which I am guessing you have already done.

**Comment by mrcoffee1026** [comment]
Yeah, I have.  Everyone was getting reminders at the same time so it wasn't using the last voted time to do anything, currently.  I've extended the remind time to a full 12 hours but I think that with the combination of the plugin not sending a reminder if they've voted in the last 24... unless they've just forgotten to vote for a full day and a half they wouldn't ever receive a reminder.  I was thinking a reminder 12 hours from the last time they voted might be acceptable since SOME of the vote sites allow use that early.  Oh well i'll look out for any other updates.  I did notice it auto reset after the month changed and that's the first time it has done that on its own.

**Comment by teozfrank** [comment]
Found the problem, the plugin was not able to check the if players voted or not when they joined, therefore it defaulted to think they didnt vote hence problem you were having. If you had checked your log files you would have seen this.

![image](https://user-images.githubusercontent.com/3265598/95018144-a67b0900-0655-11eb-8ea6-c4270c7fd4ac.png)

**Comment by teozfrank** [comment]
![image](https://user-images.githubusercontent.com/3265598/95018164-cca0a900-0655-11eb-829f-153cc2102d5c.png)

Seems its as simple as a SQL syntax error, let me fix.

**Comment by teozfrank** [comment]
Test is as follows. Join the server not having voted, deopped and have the vote reminder set to send every 30 seconds so you can see it faster.

![image](https://user-images.githubusercontent.com/3265598/95019520-c57d9900-065d-11eb-83f8-b14bee76cd45.png)

See do you get the vote reminder.

![image](https://user-images.githubusercontent.com/3265598/95019532-d75f3c00-065d-11eb-9ac7-0f3f427e6364.png)

Then vote while online on the server

![image](https://user-images.githubusercontent.com/3265598/95019551-efcf5680-065d-11eb-91f9-ae0a8cf5c30b.png)

No more prompts to vote after voting.

![image](https://user-images.githubusercontent.com/3265598/95019572-ffe73600-065d-11eb-981a-b23f59f795bb.png)

Leave server, and join server. Should not get any prompts to vote.

![image](https://user-images.githubusercontent.com/3265598/95019587-1ab9aa80-065e-11eb-96fc-c9f98a81e24a.png)

![image](https://user-images.githubusercontent.com/3265598/95019657-8ac83080-065e-11eb-889d-f5ed67a72e14.png)

Seems to work fine now. :)

**Comment by teozfrank** [comment]
new dev release for this UltimateVotes_2.0.2_SNAPSHOT-cc56edd.jar available on discord server.

**Comment by mrcoffee1026** [comment]
Yeah I have enabled this and I don't believe anyone is getting any reminders and this may be due to how you worded this change in the update posted to Spigot: "Fix vote reminding, players should be correctly reminded once every 24 hours."  If reminders aren't being given until 24 hours have past, then this time frame isn't reachable until day two:
```
    votereminder:
      enabled: true
      interval: 864000
```
The intention would be for the interval to signify when the reminder was given based on the last given vote, not 24 hours + the interval past the last given vote, since at that point it doesn't have much use.  In that case I'd want to set the interval either to 0 or to -864000 so that people might get a reminder 12 hours after they voted, as some sites allow voting every 12 hours - however, not all - and I'd only want a total of 2 reminders over the course of 24 hours with the settings I put, once at 12 hours, another at 24 hours.  It could go on indefinitely reminding every 12 hours, probably.... unless people griped about that in which case I'd reduce it to just once every 24 hours.  But since the plugin is automatically adding 24 hours to the interval, reminders can't actually be given at a time that would be useful to the player.

**Comment by teozfrank** [comment]
> 
> 
> Yeah I have enabled this and I don't believe anyone is getting any reminders and this may be due to how you worded this change in the update posted to Spigot: "Fix vote reminding, players should be correctly reminded once every 24 hours." If reminders aren't being given until 24 hours have past, then this time frame isn't reachable until day two:
> 
> ```
>     votereminder:
>       enabled: true
>       interval: 864000
> ```
> 
> The intention would be for the interval to signify when the reminder was given based on the last given vote, not 24 hours + the interval past the last given vote, since at that point it doesn't have much use. In that case I'd want to set the interval either to 0 or to -864000 so that people might get a reminder 12 hours after they voted, as some sites allow voting every 12 hours - however, not all - and I'd only want a total of 2 reminders over the course of 24 hours with the settings I put, once at 12 hours, another at 24 hours. It could go on indefinitely reminding every 12 hours, probably.... unless people griped about that in which case I'd reduce it to just once every 24 hours. But since the plugin is automatically adding 24 hours to the interval, reminders can't actually be given at a time that would be useful to the player.

No your understanding of that interval is wrong. That is when the server loops through all the players in the has not voted list and send them a vote reminder. You are over complicating how this works. Please lower that interval to 10 - 30 mins and test it out. If you see some problems with it please let me know and I can make some more tweaks but the code changes are definatly better and can be further improved with more real world testing. Players are added / removed from the voted / not voted list when they vote and are online or join the server.

**Comment by mrcoffee1026** [comment]
I have not misunderstood how it does work, you have misunderstood how it SHOULD work.  The intention of the setting is for a reminder to be displayed after the set interval has elapsed from the last vote placed from the respective player.   Since it doesn't do that, it is not properly or practically useful.  Instead you have it checking to see if the time now from the last voted time is over 24 hours before starting the interval, which is complicating the behavior into an undesirable result.

**Comment by teozfrank** [comment]
> I have not misunderstood how it does work, you have misunderstood how it SHOULD work. 

You have, you are being quite stubborn and its starting to get irritating. Stop focusing on how you think it should be coded and listen to how I have it implemeted it. 

Every single time I explain how its supposed to work you keep dismissing me and telling me no this it should be this way it has to stop NOW. I know how its supposed to work, it had a few bugs in it which are now FIXED. You changed the intervals because of the problems it had and I understand, now please change them back like I reccomended to default.

>Instead you have it checking to see if the time now from the last voted time is over 24 hours before starting the interval, which is complicating the behavior into an undesirable result.

No, thats not true have you actually seen the code? You havn't so please stop making assumptions.

Let me break it down for you in plain english how it actually works.

2 Lists.

1 list for players who voted
1 list for players who have not voted

When the plugin starts a repeating task is ran at the INTERVAL, it repeats for the entirety of when the server loads the plugin, it does not stop unless the server unloads the plugin or the server stops / restarts.

Every time that task is executed by the task scheduler the list of players who have not voted are sent a vote reminder if they are online.

How simple can it be!?

This list gets manipulated in two ways.

1. player joins, plugin checks have they voted. If they have voted they are added to the voted list, if they have not they are added to the not voted list
2. player votes they are moved from the not voted list to the voted list.

I have tested both scenarios above and they work as intended.

**Comment by teozfrank** [comment]
If you dont belive me on the timings, see for yourself. Here I have the interval set to 600 ticks which is 30 seconds.

The highlighted message is when this task gets scheduled which is part of the intialisation of the plugin when it gets loaded. 

![image](https://user-images.githubusercontent.com/3265598/97095702-b8e7d180-165a-11eb-992e-8452cd5c07d1.png)

30 seconds later it will remind players to vote and repeat every 30 seconds.

![image](https://user-images.githubusercontent.com/3265598/97095709-d74dcd00-165a-11eb-9fcf-e4616ff7e263.png)

**Comment by mrcoffee1026** [comment]
This very sentence of your explanation illustrates exactly why I'm not getting the desired behavior:
"When the plugin starts a repeating task is ran at the INTERVAL, it repeats for the entirety of when the server loads the plugin,"
So instead of having a reminder per player showing up at the Interval time past the time they voted, it's a server-wide reminder that appears to everyone... from the time the interval started based on when the server was started.  Since up till now you have repeatedly stated that players are on the "voted" list if they have voted within the last 24 hours, not if they have voted within the last interval length, I'm just never going to be able to acquire reminders from the plugin at an interval I find useful.  I'm sure the reminders are functioning the way you want them to, I just don't find that to be a useful function.  I am not arguing that they don't work the way you want them to, just stating that the way you want them to work is not useful to me, and if I put in the interval at which I want them to appear, players won't see them at all, and you've demonstrated even in your screen capture there exactly why that is the case: the interval is used from plugin start instead of from the last vote made by the player, the reminders are given to two different lists of players instead of to each player based on when they voted... and since you've said it so many times, I can assume that players are placed on the "have voted" list if they've voted within the last 24 hours.  I won't argue further on this, I'm just saying that I agree with what you said, it's doing it exactly that way, and I don't find it useful.

**Comment by mrcoffee1026** [comment]
And really, i think there should be a secondary setting for populating the voted or has not voted list.  I'm not sure how it's doing it now, but my guess is going to be the code equivalent of  "has voted since 12:00am".  I would like to think that could be altered to "has voted since 12 hours before the time right now".  But whatever, if you don't want that, then you don't.  I don't want spammy reminders so I was really hoping for something that maybe only sent a reminder twice a day per player at most, and 12 hours apart, somewhat synchronized with the last time they voted.  Only options with the current implementation are way more frequently or what it is right now which is not at all since the interval is never attained.

**Comment by teozfrank** [comment]
> So instead of having a reminder per player showing up at the Interval time past the time they voted, it's a server-wide reminder that appears to everyone..

To everyone who has not voted, yes. Its quite impractical to have the plugin keeping track live to the server tick of when they last voted and then reminding them individually. I feel that is an over complication.

> but my guess is going to be the code equivalent of "has voted since 12:00am"

No its based on 24 hours from when they have actually voted using the timestamp of when they last voted.

> I don't want spammy reminders so I was really hoping for something that maybe only sent a reminder twice a day per player at most

Thats why then you set a reminder interval thats not so spammy, maybe every 1 hour or 6 hours or something as most servers do, but it wont be in anyway syncronised to when the player last voted as its a fixed interval for the entire server.

Now what could be useful if you wanted is we could limit the amount of times a player gets a reminder in a period? For example a max of 3 reminders per 24 hours or something like that?

**Comment by mrcoffee1026** [comment]
A limit would get us closer to the behavior I'm hoping for.  What would get it even closer still would be an option for how long to look back for a last vote when populating your "has voted" list:  i'd like to have that match a long interval (of when they last voted) instead of be 24 hours (since they voted).  If that's not possible I'll have to settle for it looking spammy for a short while and then having the reminders mysteriously vanish for the people who have reached their limit?
